dbos/_core.py

```diff
@@ -267,30 +275,49 @@ def _init_workflow(
         "queue_name": queue,
         "created_at": None,
         "updated_at": None,
+        "workflow_timeout": (  # UNIX epoch timestamp of the timeout in ms
+            int((time.time() + workflow_timeout) * 1000)
+            if workflow_timeout is not None
+            else None
+        ),
     }
 
     # If we have a class name, the first arg is the instance and do not serialize
     if class_name is not None:
         inputs = {"args": inputs["args"][1:], "kwargs": inputs["kwargs"]}
 
-    wf_status = status["status"]
-    if dbos.debug_mode:
-        get_status_result = dbos._sys_db.get_workflow_status(wfid)
-        if get_status_result is None:
-            raise DBOSNonExistentWorkflowError(wfid)
-        wf_status = get_status_result["status"]
-    else:
-        # Synchronously record the status and inputs for workflows
-        # TODO: Make this transactional (and with the queue step below)
-        wf_status = dbos._sys_db.insert_workflow_status(
-            status, max_recovery_attempts=max_recovery_attempts
-        )
+    # Synchronously record the status and inputs for workflows
+    # TODO: Make this transactional (and with the queue step below)
+    wf_status, wf_timeout = dbos._sys_db.insert_workflow_status(
+        status, max_recovery_attempts=max_recovery_attempts
+    )
+
+    # TODO: Modify the inputs if they were changed by `update_workflow_inputs`
+    dbos._sys_db.update_workflow_inputs(wfid, _serialization.serialize_args(inputs))
 
-        # TODO: Modify the inputs if they were changed by `update_workflow_inputs`
-        dbos._sys_db.update_workflow_inputs(wfid, _serialization.serialize_args(inputs))
+    if queue is not None and wf_status == WorkflowStatusString.ENQUEUED.value:
+        dbos._sys_db.enqueue(wfid, queue)
+    if wf_timeout is not None:
+        evt = threading.Event()
+        dbos.stop_events.append(evt)
+
+        def timeout_func() -> None:
+            try:
+                assert wf_timeout is not None
+                time_to_wait_sec = (wf_timeout - (time.time() * 1000)) / 1000
+                if time_to_wait_sec > 0:
+                    was_stopped = evt.wait(time_to_wait_sec)
+                    if was_stopped:
+                        return
+                dbos.cancel_workflow(wfid)
```

Comment:
For now, cancel will happily mark a finished workflow as canceled. Can we make sure that cancel will not cancel any completed workflows? Maybe add some checks in `dbos.cancel_workflow()`.